lib/ecto/repo/schema.ex

```diff
@@ -379,7 +381,9 @@ defmodule Ecto.Repo.Schema do
     # On insert, we always merge the whole struct into the
     # changeset as changes, except the primary key if it is nil.
     changeset = put_repo_and_action(changeset, :insert, repo, tuplet)
-    changeset = Relation.surface_changes(changeset, struct, insertable_fields ++ assocs)
+    changeset = Relation.surface_changes(changeset, struct, surfaceables)
+    changeset = update_in(changeset.changes, &Map.take(&1, virtuals ++ surfaceables))
```

Comment:
I wonder if we should change `insertable_fields` to return a tuple with the insertable fields and the fields that need to be dropped, so we don't need to compute it here? And then this becomes a Map.drop? We only need to drop fields with writable that do not include `:insert`.

Reply:
good call :) updated